mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,693 @@
+        let serialized_block_header = binary_serialize(&block_header)?;
+        let block_size = serialized_block_size()?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return pending.
+        // As trying later might be successful.
+        if write_index < read_index {
+            if write_index + total_size > read_index {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let start = write_index;
+        let end = start + block_size;
+
+        // Check if an existing block header is present. If the block is there
+        // and if the overwrite flag is not set then we shouldn't write.
+        let result = deserialize_block(self.inner.clone(), start, end);
+        if result.is_ok() {
+            let deserialized_block = result.unwrap();
+            let should_not_overwrite = deserialized_block.inner().should_not_overwrite();
+            if should_not_overwrite {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+
+        write_data_maybe_wrap_around(
+            self.inner.clone(),
+            start,
+            end,
+            &serialized_block_header,
+            self.max_file_size,
+            should_flush,
+        )?;
+        let start = end;
+        let end = start + data_size;
+        write_data_maybe_wrap_around(
+            self.inner.clone(),
+            start,
+            end,
+            &serialized_data,
+            self.max_file_size,
+            should_flush,
+        )?;
+
+        self.pointers
+            .write
+            .store(end % self.max_file_size, Ordering::SeqCst);
+
+        self.order.store(order + 1, Ordering::SeqCst);
+
+        self.flush_state
+            .update(1, total_size, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        self.wake_up_task();
+
+        Ok(Key { offset: key as u64 })
+    }
+
+    fn batch(&self, count: usize) -> StorageResult<VecDeque<(Key, Publication)>> {
+        let write_index = self.pointers.write.load(Ordering::SeqCst);
+        let read_index = self.pointers.read.load(Ordering::SeqCst);
+        let block_size = serialized_block_size()?;
+
+        // If read would go into where writes are happening then we don't have data to read.
+        if read_index == write_index {
+            return Ok(VecDeque::new());
+        }
+
+        let mut start = read_index;
+        let mut vdata = VecDeque::new();
+        for _ in 0..count {
+            // This case shouldn't be pending, as we must have gotten something we can process.
+            if start >= write_index {
+                break;
+            }
+            let block;
```

Comment:
didn't see this before. do you know how it works? compiler converts it to something like this?

```rust
let block = { ..., binary_deserialize(bytes)? }
```